site/app/controllers/grading/ImagesController.php

```diff
@@ -177,22 +182,37 @@ public function ajaxUploadImagesFiles() {
 
     /**
      * @Route("/courses/{_semester}/{_course}/flag_user_image", methods={"POST"})
-     * @AccessControl(role="INSTRUCTOR")
+     * @AccessControl(role="FULL_ACCESS_GRADER")
      */
-    public function flagUserImage() {
+    public function flagUserImage(): JsonResponse {
+        $user_id = $_POST['user_id'];
+
         if ($_POST['flag'] === 'true') {
             $new_state = 'flagged';
-            $success_msg = $_POST['user_id'] . '\'s image was successfully flagged.';
-            $fail_msg = 'Some error occurred flagging ' . $_POST['user_id'] . '\'s image.';
+            $icon_html = ImagesView::UNDO_ICON_HTML;
+            $href = "javascript:flagUserImage('$user_id', false)";
         }
         else {
             $new_state = 'preferred';
-            $success_msg = $_POST['user_id'] . '\'s image was successfully unflagged.';
-            $fail_msg = 'Some error occurred unflagging ' . $_POST['user_id'] . '\'s image.';
+            $icon_html = ImagesView::FLAG_ICON_HTML;
+            $href = "javascript:flagUserImage('$user_id', true)";
+        }
+
+        $result = $this->core->getQueries()->updateUserDisplayImageState($user_id, $new_state);
+        $user = $this->core->getQueries()->getSubmittyUser($user_id);
+
+        if ($result) {
+            $display_image = $user->getDisplayImage();
+
+            return JsonResponse::getSuccessResponse([
+                'first_last_username' => $user->getDisplayedFirstName() . ' ' . $user->getDisplayedLastName(),
+                'image_data' => $display_image->getImageBase64MaxDimension(ImagesView::IMG_MAX_DIMENSION),
+                'image_mime_type' => $display_image->getMimeType(),
+                'icon_html' => $icon_html,
+                'href' => $href
+            ]);
         }
 
-        $result = $this->core->getQueries()->updateUserDisplayImageState($_POST['user_id'], $new_state);
-        $result ? $this->core->addSuccessMessage($success_msg) : $this->core->addErrorMessage($fail_msg);
-        return new RedirectResponse($this->core->buildCourseUrl(['student_photos']));
+        return JsonResponse::getErrorResponse('There was an error flagging or unflagging the user\'s image');
```

Comment:
can we keep using $fail_msg to show the exact msg whether it was flagging or unflagging time error

Reply:
Updated error message to indicate if the photo was being flagged vs unflagged.